File: godot-glue/src/nativeMain/kotlin/godot/internal/Jvm.kt
package godot.internal

import jni.*
import kotlinx.cinterop.*

class Jvm private constructor(private val jvm: JavaVM,
                              private val env: JNIEnv) {
    fun loadBinding() {

    }

    fun unloadBinding() {

    }

    fun destroy() {
        memScoped {
            val vm = alloc<JavaVMVar> {
                value = jvm
            }
            nullSafe(jvm.pointed.DestroyJavaVM)(vm.ptr)
        }
    }


    companion object {
        fun maybeCreate(): Jvm {
            return memScoped {
                val existingJavaVM = getExistingJavaVM()
                val env = allocPointerTo<JNIEnvVar>().ptr

                if (existingJavaVM != null) {
                    println("Re-using existing jvm ...")
                    val vm = alloc<JavaVMVar> {
                        value = existingJavaVM
                    }
                    nullSafe(existingJavaVM.pointed.GetEnv)(vm.ptr, env.reinterpret(), JNI_VERSION_1_8)
                    Jvm(existingJavaVM, nullSafe(env.pointed.value!!.pointed.value))
                } else {
                    val args = cValue<JavaVMInitArgs> {
                        version = JNI_VERSION_1_8
                        val jvmOptions = listOf(
                            "-Xcheck:jni"
                            //"-verbose:jni"
                        )
                        nOptions = jvmOptions.size
                        options = allocArray<JavaVMOption>(nOptions).also { options ->
                            jvmOptions.forEachIndexed { i, option ->
                                options[i].optionString = option.cstr.ptr
                            }
                        }
                    }
                    val vm = allocPointerTo<JavaVMVar>().ptr
                    val result = JNI_CreateJavaVM(vm, env.reinterpret(), args.ptr)
                    require(result == JNI_OK) { "Failed to create Java Virtual Machine! err=$result" }
                    Jvm(nullSafe(vm.pointed.value!!.pointed.value), nullSafe(env.pointed.value!!.pointed.value))
                }
            }
        }

        private fun getExistingJavaVM(): JavaVM? {
            return memScoped {
                val buffer = allocPointerTo<JavaVMVar>().ptr
                val vmCounts = alloc<IntVar>()
                JNI_GetCreatedJavaVMs(buffer, 1, vmCounts.ptr)
                if (vmCounts.value > 0) {
                    return buffer.pointed.value!!.pointed.value
                } else {
                    null
                }
            }

        }
    }
}

File: godot-loader2/jni_wrapper/src/jvm.cpp
#include <cassert>
#include "jvm.h"

namespace jni {
    JavaVM* Jvm::vm = nullptr;
    Env Jvm::env = Env(nullptr);

    void Jvm::init(const InitArgs& initArgs) {
        auto res = getExisting();
        if (res == nullptr) {
            res = create(initArgs);
        }
        assert(res != nullptr);
        vm = res;
        version = initArgs.version;
    }

    void Jvm::destroy() {
        vm->DetachCurrentThread();
        vm->DestroyJavaVM();
    }

    JavaVM *Jvm::create(const InitArgs& initArgs) {
        auto nOptions = initArgs.options.size();
        JavaVMOption options[nOptions];
        JavaVMInitArgs args;
        args.version = initArgs.version;
        args.nOptions = nOptions;
        args.options = options;

        for (auto i = 0; i < nOptions; i++) {
            args.options[i].optionString = (char* ) initArgs.options[i].c_str();
        }

        JavaVM* vm;
        JNIEnv* env;
        auto result = JNI_CreateJavaVM(&vm, (void**) &env, (void*) &args);
        if (result != JNI_OK) {
            throw JniError("Failed to create a new vm!");
        }
        return vm;
    }

    JavaVM* Jvm::getExisting() {
        JavaVM* buffer[1];
        int count;
        auto result = JNI_GetCreatedJavaVMs(buffer, 1, &count);
        if (result != JNI_OK) {
            throw JniError("Failed to retrieve existing vm!");
        }
        if (count > 0) {
            return buffer[0];
        }
        return nullptr;
    }

    Env &Jvm::attach() {
        if (Jvm::env.isValid()) {
            return Jvm::env;
        }
        JNIEnv* env;
        auto result = vm->GetEnv((void**) &env, version);
        if (result == JNI_EDETACHED) {
            result = vm->AttachCurrentThread((void**) &env, nullptr);
            if (result != JNI_OK) {
                throw JniError("Failed to attach vm to current thread!");
            }
        }
        Jvm::env = Env(env);
        return Jvm::env;
    }

    void Jvm::detach() {
        auto result = vm->DetachCurrentThread();
        if (result != JNI_OK) {
            throw JniError("Failed to detach vm to current thread!");
        }
        Jvm::env = Env(nullptr);
    }

    Env &Jvm::currentEnv() {
        return attach();
    }
}